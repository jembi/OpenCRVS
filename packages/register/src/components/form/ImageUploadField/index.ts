export * from '@register/components/form/ImageUploadField/ImageUploadField'
