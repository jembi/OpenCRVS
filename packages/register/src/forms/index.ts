import { Validation } from '../utils/validate'
import { FormattedMessage } from 'react-intl'
import {
  ISelectOption as SelectComponentOption,
  IRadioOption as RadioComponentOption,
  ICheckboxOption as CheckboxComponentOption
} from '@opencrvs/components/lib/forms'

export const TEXT = 'TEXT'
export const TEL = 'TEL'
export const NUMBER = 'NUMBER'
export const RADIO_GROUP = 'RADIO_GROUP'
export const CHECKBOX_GROUP = 'CHECKBOX_GROUP'
export const DATE = 'DATE'
export const TEXTAREA = 'TEXTAREA'
export const SUBSECTION = 'SUBSECTION'
export const LIST = 'LIST'
export const PARAGRAPH = 'PARAGRAPH'
export const DOCUMENTS = 'DOCUMENTS'
export const SELECT_WITH_OPTIONS = 'SELECT_WITH_OPTIONS'
export const SELECT_WITH_DYNAMIC_OPTIONS = 'SELECT_WITH_DYNAMIC_OPTIONS'
export const IMAGE_UPLOADER_WITH_OPTIONS = 'IMAGE_UPLOADER_WITH_OPTIONS'

export interface ISelectOption {
  value: SelectComponentOption['value']
  label: FormattedMessage.MessageDescriptor
}
export interface IRadioOption {
  value: RadioComponentOption['value']
  label: FormattedMessage.MessageDescriptor
}
export interface ICheckboxOption {
  value: CheckboxComponentOption['value']
  label: FormattedMessage.MessageDescriptor
}

export interface IDynamicOptions {
  dependency: string
  options: { [key: string]: ISelectOption[] }
}

export type IFormFieldValue = string | string[] | boolean | IFileValue[]

export type IFileValue = {
  optionValues: IFormFieldValue[]
  type: string
  data: string
}

export interface IFormFieldBase {
  name: string
  type: IFormField['type']
  label: FormattedMessage.MessageDescriptor
  validate: Validation[]
  required?: boolean
  prefix?: string
  postfix?: string
  disabled?: boolean
  initialValue?: IFormFieldValue
  conditionals?: IConditional[]
  description?: FormattedMessage.MessageDescriptor
}

export interface ISelectFormFieldWithOptions extends IFormFieldBase {
  type: typeof SELECT_WITH_OPTIONS
  options: ISelectOption[]
}
export interface ISelectFormFieldWithDynamicOptions extends IFormFieldBase {
  type: typeof SELECT_WITH_DYNAMIC_OPTIONS
  dynamicOptions: IDynamicOptions
}

export interface IRadioGroupFormField extends IFormFieldBase {
  type: typeof RADIO_GROUP
  options: IRadioOption[]
}

export interface ITextFormField extends IFormFieldBase {
  type: typeof TEXT
}

export interface ITelFormField extends IFormFieldBase {
  type: typeof TEL
}
export interface INumberFormField extends IFormFieldBase {
  type: typeof NUMBER
  step?: number
}
export interface ICheckboxGroupFormField extends IFormFieldBase {
  type: typeof CHECKBOX_GROUP
  options: ICheckboxOption[]
}
export interface IDateFormField extends IFormFieldBase {
  type: typeof DATE
}
export interface ITextareaFormField extends IFormFieldBase {
  type: typeof TEXTAREA
}
export interface ISubsectionFormField extends IFormFieldBase {
  type: typeof SUBSECTION
}
export interface IDocumentsFormField extends IFormFieldBase {
  type: typeof DOCUMENTS
}
export interface IListFormField extends IFormFieldBase {
  type: typeof LIST
  items: FormattedMessage.MessageDescriptor[]
}
export interface IParagraphFormField extends IFormFieldBase {
  type: typeof PARAGRAPH
}
export interface IImageUploaderWithOptionsFormField extends IFormFieldBase {
  type: typeof IMAGE_UPLOADER_WITH_OPTIONS
  optionSection: IFormSection
}

export type IFormField =
  | ITextFormField
  | ITelFormField
  | INumberFormField
  | ISelectFormFieldWithOptions
  | ISelectFormFieldWithDynamicOptions
  | IRadioGroupFormField
  | ICheckboxGroupFormField
  | IDateFormField
  | ITextareaFormField
  | ISubsectionFormField
  | IDocumentsFormField
  | IListFormField
  | IParagraphFormField
  | IImageUploaderWithOptionsFormField

export interface IConditional {
  action: string
  expression: string
}

export interface IConditionals {
  fathersDetailsExist: IConditional
  permanentAddressSameAsMother: IConditional
  addressSameAsMother: IConditional
  countryPermanent: IConditional
  statePermanent: IConditional
  districtPermanent: IConditional
  addressLine4Permanent: IConditional
  addressLine3Options1Permanent: IConditional
  country: IConditional
  state: IConditional
  district: IConditional
  addressLine4: IConditional
  addressLine3Options1: IConditional
  uploadDocForWhom: IConditional
}

export type ViewType = 'form' | 'preview' | 'review'

export interface IFormSection {
  id: string
  viewType: ViewType
  name: FormattedMessage.MessageDescriptor
  title: FormattedMessage.MessageDescriptor
  fields: IFormField[]
  disabled?: boolean
}

export interface IForm {
  sections: IFormSection[]
}

export interface Ii18nSelectOption {
  value: string
  label: string
}

export interface Ii18nFormFieldBase {
  name: string
  type: string
  label: string
  description?: string
  validate: Validation[]
  required?: boolean
  prefix?: string
  postfix?: string
  disabled?: boolean
  conditionals?: IConditional[]
}

export interface Ii18nSelectFormField extends Ii18nFormFieldBase {
  type: typeof SELECT_WITH_OPTIONS
  options: SelectComponentOption[]
}

export interface Ii18nRadioGroupFormField extends Ii18nFormFieldBase {
  type: typeof RADIO_GROUP
  options: RadioComponentOption[]
}

export interface Ii18nTextFormField extends Ii18nFormFieldBase {
  type: typeof TEXT
}
export interface Ii18nTelFormField extends Ii18nFormFieldBase {
  type: typeof TEL
}
export interface Ii18nNumberFormField extends Ii18nFormFieldBase {
  type: typeof NUMBER
  step?: number
}
export interface Ii18nCheckboxGroupFormField extends Ii18nFormFieldBase {
  type: typeof CHECKBOX_GROUP
  options: CheckboxComponentOption[]
}
export interface Ii18nDateFormField extends Ii18nFormFieldBase {
  type: typeof DATE
}
export interface Ii18nTextareaFormField extends Ii18nFormFieldBase {
  type: typeof TEXTAREA
}
export interface Ii18nSubsectionFormField extends Ii18nFormFieldBase {
  type: typeof SUBSECTION
}
export interface Ii18nDocumentsFormField extends Ii18nFormFieldBase {
  type: typeof DOCUMENTS
}
export interface Ii18nListFormField extends Ii18nFormFieldBase {
  type: typeof LIST
  items: FormattedMessage.MessageDescriptor[]
}
export interface Ii18nParagraphFormField extends Ii18nFormFieldBase {
  type: typeof PARAGRAPH
}
export interface Ii18nImageUploaderWithOptionsFormField
  extends Ii18nFormFieldBase {
  type: typeof IMAGE_UPLOADER_WITH_OPTIONS
  optionSection: IFormSection
}

export type Ii18nFormField =
  | Ii18nTextFormField
  | Ii18nTelFormField
  | Ii18nNumberFormField
  | Ii18nSelectFormField
  | Ii18nRadioGroupFormField
  | Ii18nCheckboxGroupFormField
  | Ii18nDateFormField
  | Ii18nTextareaFormField
  | Ii18nSubsectionFormField
  | Ii18nDocumentsFormField
  | Ii18nListFormField
  | Ii18nParagraphFormField
  | Ii18nImageUploaderWithOptionsFormField

export interface IFormSectionData {
  [key: string]: IFormFieldValue
}

export interface IFormData {
  [key: string]: IFormSectionData
}
