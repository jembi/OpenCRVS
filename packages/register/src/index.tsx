import * as React from 'react'
import * as ReactDOM from 'react-dom'
import { injectGlobal } from '@register/styledComponents'
import { App } from '@register/App'
import registerServiceWorker from '@register/registerServiceWorker'
import { createStore } from '@register/store'
import * as actions from '@register/notification/actions'
import { storage } from '@register/storage'
import * as Sentry from '@sentry/browser'
import * as LogRocket from 'logrocket'
import { SubmissionController } from '@register/SubmissionController'

storage.configStorage('OpenCRVS')

// Injecting global styles for the body tag - used only once
// eslint-disable-line
injectGlobal`
  body {
    margin: 0;
    padding: 0;
    overflow-y:overlay;
  }
`

const { store, history } = createStore()

if (
  window.location.hostname !== 'localhost' &&
  window.location.hostname !== '127.0.0.1'
) {
  // setup error reporting using sentry
  Sentry.init({
    dsn: 'https://8f6ba426b20045f1b91528d5fdc214b5@sentry.io/1401900'
  })

  // setup log rocket to ship log messages and record user errors
  LogRocket.init('hxf1hb/opencrvs')

  // Integrate the two
  Sentry.configureScope(scope => {
    scope.addEventProcessor(async event => {
      if (!event.extra) {
        event.extra = {}
      }
      const sessionUrl = await new Promise(resolve => {
        LogRocket.getSessionURL(url => {
          resolve(url)
        })
      })
      event.extra.sessionURL = sessionUrl
      return event
    })
  })
}

function onNewConentAvailable(waitingSW: ServiceWorker | null) {
  if (waitingSW) {
    const action = actions.showNewContentAvailableNotification(waitingSW)
    store.dispatch(action)
  }
}

function onBackGroundSync() {
  const channel = new BroadcastChannel(
    window.config.BACKGROUND_SYNC_BROADCAST_CHANNEL
  )
  channel.onmessage = e => {
    const syncCount = typeof e.data === 'number' ? e.data : 0
    const action = actions.showBackgroundSyncedNotification(syncCount)
    store.dispatch(action)
  }
}

onBackGroundSync()

ReactDOM.render(
  <App store={store} history={history} />,
  document.getElementById('root')
)

registerServiceWorker(onNewConentAvailable)
new SubmissionController(store).start()
