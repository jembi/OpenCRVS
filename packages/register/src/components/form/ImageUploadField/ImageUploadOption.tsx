import * as React from 'react'
import styled from '@register/styledComponents'
import { ActionPage, Box } from '@opencrvs/components/lib/interface'
import { ImageUploader } from '@opencrvs/components/lib/forms'
import { ArrowForward } from '@opencrvs/components/lib/icons'
import { FormFieldGenerator } from '@register/components/form'
import { IFormSection, IFormSectionData, IFileValue } from '@register/forms'
import { hasFormError } from '@register/forms/utils'
import { BodyContent } from '@opencrvs/components/lib/layout'
import * as Jimp from 'jimp'
import { ALLOWED_IMAGE_TYPE } from '@register/utils/constants'
import { defineMessages, InjectedIntlProps, injectIntl } from 'react-intl'

const FormContainer = styled.div`
  padding: 35px 25px;
`
export const OverlayContainer = styled.div`
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background: ${({ theme }) => theme.colors.background};
  z-index: 4;
  width: 100%;
  height: 100%;
  overflow-y: scroll;
`
const FormAction = styled.div`
  display: flex;
  justify-content: center;
`
const FormImageUploader = styled(ImageUploader)`
  box-shadow: 0 0 13px 0 rgba(0, 0, 0, 0.27);
`
const UploadErrorSec = styled.div`
  display: flex;
  justify-content: center;
  color: ${({ theme }) => theme.colors.error};
  margin-top: 10px;
`

const messages: {
  [key: string]: ReactIntl.FormattedMessage.MessageDescriptor
} = defineMessages({
  uploadError: {
    id: 'imageUploadOption.upload.error',
    defaultMessage: 'Must be in JPEG/JPG/PNG format',
    description: 'Show error messages while uploading'
  }
})

type State = {
  data: IFormSectionData
  showUploadButton: boolean
  errorOnUpload: boolean
}

const getBase64String = (file: File) => {
  return new Promise((resolve, reject) => {
    const reader = new FileReader()
    reader.readAsDataURL(file)
    reader.onload = () => {
      if (reader.result) {
        return resolve(reader.result)
      }
    }
    reader.onerror = error => reject(error)
  })
}

type IProps = {
  option: IFormSection
  backLabel?: string
  title?: string
  onComplete: (file: IFileValue) => void
  toggleNestedSection: () => void
}

type IFullProps = InjectedIntlProps & IProps

export class ImageUploadOptionClass extends React.Component<IFullProps, State> {
  constructor(props: IFullProps) {
    super(props)
    this.state = {
      data: {},
      showUploadButton: false,
      errorOnUpload: false
    }
  }

  storeData = (documentData: IFormSectionData) => {
    const showUploadButton = this.shouldShowUploadButton(documentData)
    this.setState({
      data: documentData,
      showUploadButton
    })
  }

  shouldShowUploadButton = (documentData: IFormSectionData) => {
    return documentData && !hasFormError(this.props.option.fields, documentData)
  }

  handleFileChange = (uploadedImage: File) => {
    getBase64String(uploadedImage).then(data => {
      let base64String = data as string
      base64String = base64String.split('base64,')[1]
      Jimp.read(new Buffer(base64String, 'base64'))
        .then(buffer => {
          if (!ALLOWED_IMAGE_TYPE.includes(buffer.getMIME())) {
            throw new Error('File type not supported')
          } else if (uploadedImage.size > 2097152) {
            return buffer
              .resize(2000, Jimp.AUTO)
              .quality(70)
              .getBase64Async(buffer.getMIME())
          }
          return data as string
        })
        .then(buffer => {
          this.props.onComplete({
            optionValues: Object.values(this.state.data),
            type: uploadedImage.type,
            data: buffer
          })
        })
        .catch(() => {
          this.setState({
            errorOnUpload: true
          })
        })
    })
  }

  render = () => {
    const { option, title, backLabel, intl } = this.props
    return (
      <OverlayContainer>
        <ActionPage
          title={title ? title : 'Upload'}
          backLabel={backLabel}
          goBack={this.props.toggleNestedSection}
        >
          <BodyContent>
            <FormContainer>
              <Box>
                <FormFieldGenerator
                  id={option.id}
                  onChange={this.storeData}
                  setAllFieldsDirty={false}
                  fields={option.fields}
                />
                {this.state.showUploadButton && (
                  <FormAction>
                    <FormImageUploader
                      id="upload_document"
                      title={title ? title : 'Upload'}
                      icon={() => <ArrowForward />}
                      handleFileChange={this.handleFileChange}
                    />
                  </FormAction>
                )}
                {this.state.showUploadButton && this.state.errorOnUpload && (
                  <UploadErrorSec id="upload-error-sec">
                    {intl.formatMessage(messages.uploadError)}
                  </UploadErrorSec>
                )}
              </Box>
            </FormContainer>
          </BodyContent>
        </ActionPage>
      </OverlayContainer>
    )
  }
}
export const ImageUploadOption = injectIntl<IFullProps>(ImageUploadOptionClass)
