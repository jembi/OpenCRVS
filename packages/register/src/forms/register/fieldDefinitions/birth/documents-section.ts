import { defineMessages } from 'react-intl'
import {
  IFormSection,
  PARAGRAPH,
  IMAGE_UPLOADER_WITH_OPTIONS,
  RADIO_GROUP,
  SELECT_WITH_DYNAMIC_OPTIONS,
  DYNAMIC_LIST
} from 'src/forms'
import { birthFieldToAttachmentTransformer } from './mappings/mutation/documents-mappings'
import { birthAttachmentToFieldTransformer } from './mappings/query/documents-mappings'
import { diffDoB } from 'src/forms/utils'

const messages = defineMessages({
  documentsTab: {
    id: 'register.form.tabs.documentsTab',
    defaultMessage: 'Documents',
    description: 'Tab title for Documents'
  },
  documentsTitle: {
    id: 'register.form.section.documentsTitle',
    defaultMessage: 'Supporting documents',
    description: 'Form section title for Documents'
  },
  uploadImage: {
    id: 'register.form.section.documents.uploadImage',
    defaultMessage: 'Upload a photo of the supporting document',
    description: 'Title for the upload image button'
  },
  paragraph: {
    id: 'register.form.section.documents.paragraph',
    defaultMessage:
      'For birth regiatration of children below 5 years old, one of the documents listed below is required:',
    description: 'Documents Paragraph text'
  },
  informantAttestation: {
    id: 'register.form.section.documents.list.informantAttestation',
    defaultMessage: 'Attestation of the informant, or',
    description: 'Attested document of the informant'
  },
  attestedVaccination: {
    id: 'register.form.section.documents.list.attestedVaccination',
    defaultMessage: 'Attested copy of the vaccination (EPI) card, or',
    description: 'Attested copy of the vaccination card'
  },
  attestedBirthRecord: {
    id: 'register.form.section.documents.list.attestedBirthRecord',
    defaultMessage: 'Attested copy of hospital document or birth record, or',
    description: 'Attested copy of hospital document'
  },
  certification: {
    id: 'register.form.section.documents.list.certification',
    defaultMessage:
      'Certification regarding NGO worker authorized by registrar in favour of date of birth, or',
    description: 'Certification regarding NGO worker'
  },
  otherDocuments: {
    id: 'register.form.section.documents.list.otherDocuments',
    defaultMessage:
      'Attested copy(s) of the document as prescribed by the Registrar',
    description: 'Attested copy(s) of the document'
  },
  documentsUploadName: {
    id: 'register.form.section.upload.documentsName',
    defaultMessage: 'Documents Upload',
    description: 'Tab title for Documents Upload'
  },
  documentsUploadTitle: {
    id: 'register.form.section.upload.documentsTitle',
    defaultMessage: 'Supporting documents',
    description: 'Form section title for Documents'
  },
  uploadDocForWhom: {
    id: 'formFields.uploadDocForWhom',
    defaultMessage: 'Whose suppoting document are you uploading?',
    description: 'Question to ask user, for whom, documents are being uploaded'
  },
  uploadDocForChild: {
    id: 'formFields.uploadDocForChild',
    defaultMessage: 'Child',
    description: 'Label for radio option Child'
  },
  uploadDocForMother: {
    id: 'formFields.uploadDocForMother',
    defaultMessage: 'Mother',
    description: 'Label for radio option Mother'
  },
  uploadDocForFather: {
    id: 'formFields.uploadDocForFather',
    defaultMessage: 'Father',
    description: 'Label for radio option Father'
  },
  uploadDocForOther: {
    id: 'formFields.uploadDocForOther',
    defaultMessage: 'Other',
    description: 'Label for radio option Other'
  },
  whatDocToUpload: {
    id: 'formFields.whatDocToUpload',
    defaultMessage: 'Which document type are you uploading?',
    description:
      'Question to ask user, what type of documents are being uploaded'
  },
  docTypeBR: {
    id: 'formFields.docTypeBR',
    defaultMessage: 'Birth Registration',
    description: 'Label for radio option Birth Registration'
  },
  docTypeNIDFront: {
    id: 'formFields.docTypeNIDFront',
    defaultMessage: 'National ID (front)',
    description: 'Label for select option radio option NID front'
  },
  docTypeNIDBack: {
    id: 'formFields.docTypeNIDBack',
    defaultMessage: 'National ID (back)',
    description: 'Label for select option radio option NID back'
  },
  docTypePassport: {
    id: 'formFields.docTypePassport',
    defaultMessage: 'Passport',
    description: 'Label for radio option Passport'
  },
  docTypeSC: {
    id: 'formFields.docTypeSC',
    defaultMessage: 'School Certificate',
    description: 'Label for radio option School Certificate'
  },
  docTypeOther: {
    id: 'formFields.docTypeOther',
    defaultMessage: 'Other',
    description: 'Label for radio option Other'
  },
  docTypeChildBirthProof: {
    id: 'formFields.docTypeChildBirthProof',
    defaultMessage: 'Proof of Place and Date of Birth',
    description: 'Label for select option Child Birth Proof'
  },
  docTypeEPICard: {
    id: 'formFields.docTypeEPICard',
    defaultMessage: 'EPI Card',
    description: 'Label for select option EPI Card'
  },
  docTypeDoctorCertificate: {
    id: 'formFields.docTypeDoctorCertificate',
    defaultMessage: 'Doctor Certificate',
    description: 'Label for select option Doctor Certificate'
  },
  proofOfMothersID: {
    id: 'formFields.proofOfMothersID',
    defaultMessage: "Proof of Mother's ID",
    description: 'Label for list item Mother ID Proof'
  },
  proofOfFathersID: {
    id: 'formFields.proofOfFathersID',
    defaultMessage: "Proof of Father's ID",
    description: 'Label for list item Father ID Proof'
  },
  proofOfBirthPlaceAndDate: {
    id: 'formFields.proofOfBirthPlaceAndDate',
    defaultMessage: 'Proof of Place and Date of Birth of Child',
    description: 'Label for list item Child Birth Proof'
  },
  proofOfParentPermanentAddress: {
    id: 'formFields.proofOfParentPermanentAddress',
    defaultMessage: 'Proof of Permanent Address of Parent',
    description: 'Label for list item Parent Permanent Address Proof'
  },
  proofOfEPICardOfChild: {
    id: 'formFields.proofOfEPICardOfChild',
    defaultMessage: 'EPI Card of Child',
    description: 'Label for list item EPI Card of Child'
  },
  proofOfDocCertificateOfChild: {
    id: 'formFields.proofOfDocCertificateOfChild',
    defaultMessage:
      "Certificate from doctor to prove child's age OR School certificate",
    description: 'Label for list item Doctor Certificate'
  }
})

export const documentsSection: IFormSection = {
  id: 'documents',
  viewType: 'form',
  name: messages.documentsTab,
  title: messages.documentsTitle,
  fields: [
    {
      name: 'image_uploader',
      type: IMAGE_UPLOADER_WITH_OPTIONS,
      label: messages.uploadImage,
      initialValue: '',
      validate: [],
      optionSection: {
        id: 'documents-upload',
        viewType: 'form',
        name: messages.documentsUploadName,
        title: messages.documentsUploadTitle,
        fields: [
          {
            name: 'uploadDocForWhom',
            type: RADIO_GROUP,
            label: messages.uploadDocForWhom,
            required: true,
            initialValue: '',
            hideAsterisk: true,
            validate: [],
            options: [
              { value: 'Child', label: messages.uploadDocForChild },
              { value: 'Mother', label: messages.uploadDocForMother },
              { value: 'Father', label: messages.uploadDocForFather },
              { value: 'Other', label: messages.uploadDocForOther }
            ]
          },
          {
            name: 'whatDocToUpload',
            type: SELECT_WITH_DYNAMIC_OPTIONS,
            label: messages.whatDocToUpload,
            required: true,
            hideAsterisk: true,
            initialValue: '',
            validate: [],
            dynamicOptions: {
              dependency: 'uploadDocForWhom',
              options: {
                Child: [
                  {
                    value: 'Proof of Place and Date of Birth',
                    label: messages.docTypeChildBirthProof
                  },
                  { value: 'EPI Card', label: messages.docTypeEPICard },
                  {
                    value: 'Doctor Certificate',
                    label: messages.docTypeDoctorCertificate
                  },
                  { value: 'School Certificate', label: messages.docTypeSC }
                ],
                Mother: [
                  { value: 'Birth Registration', label: messages.docTypeBR },
                  {
                    value: 'National ID (front)',
                    label: messages.docTypeNIDFront
                  },
                  {
                    value: 'National ID (back)',
                    label: messages.docTypeNIDBack
                  },
                  { value: 'Passport', label: messages.docTypePassport },
                  { value: 'School Certificate', label: messages.docTypeSC },
                  { value: 'Other', label: messages.docTypeOther }
                ],
                Father: [
                  { value: 'Birth Registration', label: messages.docTypeBR },
                  {
                    value: 'National ID (front)',
                    label: messages.docTypeNIDFront
                  },
                  {
                    value: 'National ID (back)',
                    label: messages.docTypeNIDBack
                  },
                  { value: 'Passport', label: messages.docTypePassport },
                  { value: 'School Certificate', label: messages.docTypeSC },
                  { value: 'Other', label: messages.docTypeOther }
                ],
                Other: [
                  { value: 'Birth Registration', label: messages.docTypeBR },
                  {
                    value: 'National ID (front)',
                    label: messages.docTypeNIDFront
                  },
                  {
                    value: 'National ID (back)',
                    label: messages.docTypeNIDBack
                  },
                  { value: 'Passport', label: messages.docTypePassport },
                  { value: 'School Certificate', label: messages.docTypeSC },
                  { value: 'Other', label: messages.docTypeOther }
                ]
              }
            }
          }
        ]
      },
      mapping: {
        mutation: birthFieldToAttachmentTransformer,
        query: birthAttachmentToFieldTransformer
      }
    },
    {
      name: 'paragraph',
      type: PARAGRAPH,
      label: messages.paragraph,
      initialValue: '',
      validate: []
    },
    {
      name: 'list',
      type: DYNAMIC_LIST,
      label: messages.documentsTab,
      validate: [],
      initialValue: '',
      dynamicItems: {
        dependency: 'child.childBirthDate',
        valueMapper: diffDoB,
        items: {
          within45days: [
            messages.proofOfMothersID,
            messages.proofOfFathersID,
            messages.proofOfBirthPlaceAndDate,
            messages.proofOfParentPermanentAddress
          ],
          between46daysTo5yrs: [
            messages.proofOfMothersID,
            messages.proofOfFathersID,
            messages.proofOfBirthPlaceAndDate,
            messages.proofOfParentPermanentAddress,
            messages.proofOfEPICardOfChild
          ],
          after5yrs: [
            messages.proofOfMothersID,
            messages.proofOfFathersID,
            messages.proofOfBirthPlaceAndDate,
            messages.proofOfParentPermanentAddress,
            messages.proofOfDocCertificateOfChild
          ]
        }
      }
    }
  ]
}
