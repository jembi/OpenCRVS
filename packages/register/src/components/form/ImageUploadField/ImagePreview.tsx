import * as React from 'react'
import styled from '@register/styledComponents'
import { IFileValue } from '@register/forms'
import { ActionPage } from '@opencrvs/components/lib/interface'
import { OverlayContainer } from '@register/components/form/ImageUploadField/ImageUploadOption'

const PreviewContainer = styled(OverlayContainer)`
  background: ${({ theme }) => theme.colors.black};
`
const ImageHolder = styled.div`
  position: relative;
  width: 100%;
  height: calc(100vh - 90px);
`
const Image = styled.img`
  position: absolute;
  max-width: 100vw;
  max-height: 100%;
  left: 50%;
  top: 50%;
  transform: translateX(-50%) translateY(-50%);
`

type IProps = {
  previewImage: IFileValue
  backLabel?: string
  title?: string
  goBack: () => void
}

export class ImagePreview extends React.Component<IProps> {
  render = () => {
    const { previewImage, title, backLabel, goBack } = this.props
    return (
      <PreviewContainer>
        <ActionPage
          title={title ? title : 'Preview'}
          backLabel={backLabel}
          goBack={goBack}
        >
          <ImageHolder>
            <Image id="preview_image_field" src={previewImage.data} />
          </ImageHolder>
        </ActionPage>
      </PreviewContainer>
    )
  }
}
