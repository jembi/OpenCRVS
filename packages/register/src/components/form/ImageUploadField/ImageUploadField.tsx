import * as React from 'react'
import { defineMessages, InjectedIntlProps, injectIntl } from 'react-intl'
import styled from '@register/styledComponents'
import { IconAction } from '@opencrvs/components/lib/buttons'
import { Camera } from '@opencrvs/components/lib/icons'
import { IActionProps } from '@opencrvs/components/lib/buttons/Action'
import { ActionTitle } from '@opencrvs/components/lib/buttons/IconAction'
import { FileItem } from '@opencrvs/components/lib/files'
import { IFormSection, IFileValue } from '@register/forms'
import { ImageUploadOption } from '@register/components/form/ImageUploadField/ImageUploadOption'
import { ImagePreview } from '@register/components/form/ImageUploadField/ImagePreview'

const Container = styled.div`
  width: 100%;
  border-bottom: 1px solid ${({ theme }) => theme.colors.background};
  padding-bottom: 35px;
`
const StyledIcon = styled(Camera)`
  border-radius: 2px;
  box-shadow: 0 0 4px 3px rgba(0, 0, 0, 0.1);
  height: 50px;
  width: 50px;
  background-color: ${({ theme }) => theme.colors.secondary};
`
const PhotoIconAction = styled(IconAction)`
  outline-style: dashed;
  outline-color: ${({ theme }) => theme.colors.secondary};
  outline-width: 1px;
  min-height: 90px;
  /* stylelint-disable */
  ${ActionTitle} {
    /* stylelint-enable */
    ${({ theme }) => theme.fonts.bigBodyStyle};
    margin-left: 75px;
  }
`
const FileViewer = styled.div`
  margin-top: 15px;
`
const FileViewerLabel = styled.label`
  color: ${({ theme }) => theme.colors.primary};
  ${({ theme }) => theme.fonts.bodyStyle};
`
const FileItemContainer = styled.div`
  margin-top: 12px;
`
const messages: {
  [key: string]: ReactIntl.FormattedMessage.MessageDescriptor
} = defineMessages({
  back: {
    id: 'menu.back',
    defaultMessage: 'Back',
    description: 'Back button in the menu'
  },
  uploadedList: {
    id: 'formFields.imageUpload.uploadedList',
    defaultMessage: 'Uploaded:',
    description: 'label for uploaded list'
  },
  delete: {
    id: 'formFields.imageUpload.delete',
    defaultMessage: 'Delete',
    description: 'label for delete a uploaded item'
  },
  preview: {
    id: 'formFields.imageUpload.preview',
    defaultMessage: 'Preview',
    description: 'label for preview a uploaded item'
  },
  upload: {
    id: 'register.form.upload',
    defaultMessage: 'Upload',
    description: 'title for option view'
  }
})

type IProps = {
  files?: IFileValue[]
  optionSection: IFormSection
  onComplete: (files: IFileValue[]) => void
}
type IFullProps = IActionProps & InjectedIntlProps & IProps

interface IFullFileValues extends IFileValue {
  title: string
  description: string
}

interface ITempFullFileValues extends IFileValue {
  title?: string
  description?: string
}
/* feels weired may need to change a bit */
function augmentFile(file: IFileValue): IFullFileValues {
  const augmentedFile: ITempFullFileValues = file
  if (file.optionValues) {
    augmentedFile.title = file.optionValues[0].toString()
    if (file.optionValues.length > 1) {
      augmentedFile.description = file.optionValues[1].toString()
    }
  }
  return augmentedFile as IFullFileValues
}
class ImageUploadComponent extends React.Component<
  IFullProps,
  {
    showNestedOptionSection: boolean
    previewImage: IFileValue | null
  }
> {
  constructor(props: IFullProps) {
    super(props)
    this.state = {
      showNestedOptionSection: false,
      previewImage: null
    }
  }

  toggleNestedSection = () => {
    this.setState({
      showNestedOptionSection: !this.state.showNestedOptionSection
    })
  }

  closePreviewSection = () => {
    this.setState({ previewImage: null })
  }

  onDelete = (index: number): void => {
    const files = this.props.files ? this.props.files : []
    files.splice(index, 1)
    this.props.onComplete(files)
  }

  onPreview = (file: IFileValue): void => {
    this.setState({ previewImage: file })
  }

  onComplete = (file: IFileValue) => {
    const files = this.props.files ? this.props.files : []
    this.props.onComplete(files.concat(file))
    this.toggleNestedSection()
  }

  render = () => {
    const { title, optionSection, files, intl } = this.props
    const fileList =
      files &&
      files.map((file: IFileValue, index: number) => {
        const augmentedFile: IFullFileValues = augmentFile(file)
        return (
          <FileItemContainer key={index}>
            <FileItem
              id={`file_item_${index}`}
              file={augmentedFile}
              deleteLabel={intl.formatMessage(messages.delete)}
              onDelete={() => this.onDelete(index)}
              previewLabel={intl.formatMessage(messages.preview)}
              onPreview={() => this.onPreview(file)}
            />
          </FileItemContainer>
        )
      })
    return (
      <Container>
        <PhotoIconAction
          type="button"
          icon={() => <StyledIcon />}
          title={title}
          {...this.props}
          onClick={this.toggleNestedSection}
        />

        {fileList && fileList.length > 0 && (
          <FileViewer id="file_list_viewer">
            <FileViewerLabel>
              {intl.formatMessage(messages.uploadedList)}
            </FileViewerLabel>
            {fileList}
          </FileViewer>
        )}

        {this.state.showNestedOptionSection && (
          <ImageUploadOption
            option={optionSection}
            title={intl.formatMessage(messages.upload)}
            backLabel={intl.formatMessage(messages.back)}
            onComplete={this.onComplete}
            toggleNestedSection={this.toggleNestedSection}
          />
        )}

        {this.state.previewImage && (
          <ImagePreview
            previewImage={this.state.previewImage}
            title={intl.formatMessage(messages.preview)}
            backLabel={intl.formatMessage(messages.back)}
            goBack={this.closePreviewSection}
          />
        )}
      </Container>
    )
  }
}
export const ImageUploadField = injectIntl<IFullProps>(ImageUploadComponent)
