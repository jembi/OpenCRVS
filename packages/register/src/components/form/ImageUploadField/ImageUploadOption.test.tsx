import * as React from 'react'
import { createShallowRenderedComponent } from '@register/tests/util'
import { ImageUploadOption } from '@register/components/form/ImageUploadField/ImageUploadOption'
import { documentsSection } from '@register/forms/register/fieldDefinitions/birth/documents-section'
import { IImageUploaderWithOptionsFormField } from '@opencrvs/register/src/forms'
import { IntlProvider } from 'react-intl'

describe('image upload component', () => {
  const mock = jest.fn()
  const testComponent = createShallowRenderedComponent(
    <IntlProvider locale="en">
      <ImageUploadOption
        option={
          (documentsSection.fields[0] as IImageUploaderWithOptionsFormField)
            .optionSection
        }
        onComplete={mock}
        toggleNestedSection={mock}
      />
    </IntlProvider>
  )

  it('renders without crashing', () => {
    expect(testComponent).toMatchSnapshot()
  })
})
