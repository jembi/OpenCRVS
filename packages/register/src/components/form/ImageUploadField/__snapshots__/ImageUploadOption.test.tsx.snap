// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`image upload component renders without crashing 1`] = `ShallowWrapper {}`;
